python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py

```diff
@@ -0,0 +1,142 @@
+
+        Returns:
+            Mapping[str, Any]: asset-specific metadata dictionary
+
+        """
+        dlt_base_metadata_types = {
+            "first_run",
+            "started_at",
+            "finished_at",
+            "dataset_name",
+            "destination_name",
+            "destination_type",
+        }
+
+        load_info_dict = self._cast_load_info_metadata(load_info.asdict())
+
+        # shared metadata that is displayed for all assets
+        base_metadata = {k: v for k, v in load_info_dict.items() if k in dlt_base_metadata_types}
+
+        # job metadata for specific target `resource.table_name`
+        base_metadata["jobs"] = [
+            job
+            for load_package in load_info_dict.get("load_packages", [])
+            for job in load_package.get("jobs", [])
+            if job.get("table_name") == resource.table_name
+        ]
+
+        return base_metadata
+
+    @public
+    def run(
+        self,
+        context: Union[OpExecutionContext, AssetExecutionContext],
+        **kwargs,
+    ) -> Iterator[Union[AssetMaterialization, MaterializeResult]]:
+        """Runs dlt pipeline.
+
+        Args:
+            context (Union[OpExecutionContext, AssetExecutionContext]) execution context
+            **kwargs (dict[str, Any]) keyword args passed to pipeline `run` method
+
+        Returns:
+            Iterator[Union[AssetMaterialization, MaterializeResult]] materialization in asset or op
+
+        """
+        metadata_by_key = context.assets_def.metadata_by_key
+        first_asset_metadata = next(iter(metadata_by_key.values()))
+
+        dlt_source: Optional[DltSource] = first_asset_metadata.get(META_KEY_SOURCE)
+        dlt_pipeline: Optional[Pipeline] = first_asset_metadata.get(META_KEY_PIPELINE)
+
+        if not dlt_source or not dlt_pipeline:
+            raise Exception(
+                "%s and %s must be defined on AssetSpec metadata",
+                META_KEY_SOURCE,
+                META_KEY_PIPELINE,
+            )
+
+        # Mapping of asset keys to dlt resources
+        asset_key_dlt_source_resource_mapping = zip(
+            context.selected_asset_keys, dlt_source.resources.values()
+        )
+
+        # Filter sources by asset key sub-selection
+        if context.is_subset:
+            asset_key_dlt_source_resource_mapping = [
+                (asset_key, dlt_source_resource)
+                for (asset_key, dlt_source_resource) in asset_key_dlt_source_resource_mapping
+                if asset_key in context.selected_asset_keys
+            ]
+            dlt_source = dlt_source.with_resources(
+                *[
+                    dlt_source_resource.name
+                    for (_, dlt_source_resource) in asset_key_dlt_source_resource_mapping
+                ]
+            )
+
+        load_info = dlt_pipeline.run(dlt_source, **kwargs)
+
+        for asset_key, dlt_source_resource in asset_key_dlt_source_resource_mapping:
```

Comment:
We need to inspect this pipeline for success/failure before yielding events

Reply:
To my knowledge the pipeline will throw an error if any failure is to occur. But this is something I can look into more detail.